samza-core/src/main/java/org/apache/samza/table/remote/RemoteTableProvider.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote;
+
+import java.io.ByteArrayInputStream;
+import java.io.ObjectInputStream;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.JavaTableConfig;
+import org.apache.samza.container.SamzaContainerContext;
+import org.apache.samza.operators.KV;
+import org.apache.samza.table.Table;
+import org.apache.samza.table.TableProvider;
+import org.apache.samza.table.TableSpec;
+import org.apache.samza.task.TaskContext;
+import org.apache.samza.util.RateLimiter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Provide for remote table instances
+ */
+public class RemoteTableProvider implements TableProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteTableProvider.class);
+
+  static final String READ_FN = "io.readFn";
+  static final String WRITE_FN = "io.writeFn";
+  static final String RATE_LIMITER = "io.ratelimiter";
+  static final String READ_CREDIT_FN = "io.readCreditFn";
+  static final String WRITE_CREDIT_FN = "io.writeCreditFn";
+
+  private final TableSpec tableSpec;
+  private final boolean readOnly;
+  private final List<RemoteReadableTable<?, ?>> tables = new ArrayList<>();
+  private SamzaContainerContext containerContext;
+  private TaskContext taskContext;
+
+  public RemoteTableProvider(TableSpec tableSpec) {
+    this.tableSpec = tableSpec;
+    readOnly = !tableSpec.getConfig().containsKey(WRITE_FN);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void init(SamzaContainerContext containerContext, TaskContext taskContext) {
+    this.containerContext = containerContext;
+    this.taskContext = taskContext;
+  }
+
+  private <T> T deserializeObject(String key) {
```

Comment:
Organize methods as follows:

init()
getTable()
generateConfig()
close()

private helper methods in the class